File: go-sdk/webhooks/webhooks_test.go
package webhooks

import (
	"crypto/sha256"
	"fmt"
	"reflect"
	"strings"
	"testing"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/pvium/sdks/go-sdk/models"
)

func signWebhookToken(t *testing.T, secret string, claims jwt.MapClaims) string {
	t.Helper()

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	signed, err := token.SignedString([]byte(secret))
	if err != nil {
		t.Fatalf("sign token: %v", err)
	}
	return signed
}

func TestVerifyPviumWebhookTokenVerifiesBackendHS256WebhookJWTs(t *testing.T) {
	t.Parallel()

	token := signWebhookToken(t, "webhook_secret", jwt.MapClaims{
		"event": "oauth.invite.accepted",
		"data":  map[string]any{"githubLogin": "octocat"},
		"iat":   1700000000,
		"exp":   4000000000,
	})

	payload, err := VerifyPviumWebhookToken(token, "webhook_secret", &models.VerifyPviumWebhookTokenOptions{
		ExpectedEvent: "oauth.invite.accepted",
	})
	if err != nil {
		t.Fatalf("verify webhook token: %v", err)
	}

	if payload.Event != "oauth.invite.accepted" {
		t.Fatalf("event mismatch: got %s want %s", payload.Event, "oauth.invite.accepted")
	}
	wantData := map[string]any{"githubLogin": "octocat"}
	if !reflect.DeepEqual(payload.Data, wantData) {
		t.Fatalf("data mismatch:\ngot  %#v\nwant %#v", payload.Data, wantData)
	}
}

func TestVerifyPviumWebhookTokenSupportsBackendTokensSignedWithSHA256Secret(t *testing.T) {
	t.Parallel()

	secret := "secret_abc123"
	hashed := sha256.Sum256([]byte(secret))
	token := signWebhookToken(t, fmt.Sprintf("%x", hashed), jwt.MapClaims{
		"event": "batch.payee.added",
		"data":  map[string]any{"batch": map[string]any{"id": "batch_123"}},
		"exp":   4000000000,
	})

	payload, err := VerifyPviumWebhookToken(token, secret, nil)
	if err != nil {
		t.Fatalf("verify token with hashed-secret fallback: %v", err)
	}

	if payload.Event != "batch.payee.added" {
		t.Fatalf("event mismatch: got %s want %s", payload.Event, "batch.payee.added")
	}
	wantData := map[string]any{"batch": map[string]any{"id": "batch_123"}}
	if !reflect.DeepEqual(payload.Data, wantData) {
		t.Fatalf("data mismatch:\ngot  %#v\nwant %#v", payload.Data, wantData)
	}
}

func TestResolvePviumWebhookPayloadReturnsTokenDataAndChecksBodyEvent(t *testing.T) {
	t.Parallel()

	token := signWebhookToken(t, "webhook_secret", jwt.MapClaims{
		"event": "invoice.paid",
		"data":  map[string]any{"invoiceId": "inv_123"},
		"exp":   4000000000,
	})

	resolved, err := ResolvePviumWebhookPayload(map[string]any{
		"event": "invoice.paid",
		"token": token,
	}, "webhook_secret")
	if err != nil {
		t.Fatalf("resolve webhook payload: %v", err)
	}

	if resolved.Event != "invoice.paid" {
		t.Fatalf("event mismatch: got %s want %s", resolved.Event, "invoice.paid")
	}
	wantData := map[string]any{"invoiceId": "inv_123"}
	if !reflect.DeepEqual(resolved.Data, wantData) {
		t.Fatalf("data mismatch:\ngot  %#v\nwant %#v", resolved.Data, wantData)
	}
}

func TestVerifyPviumWebhookTokenRejectsExpiredTokens(t *testing.T) {
	t.Parallel()

	token := signWebhookToken(t, "webhook_secret", jwt.MapClaims{
		"event": "invoice.paid",
		"data":  map[string]any{},
		"exp":   1700000000,
	})

	_, err := VerifyPviumWebhookToken(token, "webhook_secret", &models.VerifyPviumWebhookTokenOptions{
		Now: time.Unix(1800000000, 0),
	})
	if err == nil {
		t.Fatal("expected expired token error")
	}
	if !strings.Contains(err.Error(), "Expired Pvium webhook token") {
		t.Fatalf("error mismatch: got %q", err.Error())
	}
}

func TestVerifyPviumWebhookTokenCanDisableHashedSecretFallback(t *testing.T) {
	t.Parallel()

	secret := "secret_abc123"
	hashed := sha256.Sum256([]byte(secret))
	token := signWebhookToken(t, fmt.Sprintf("%x", hashed), jwt.MapClaims{
		"event": "oauth.invite.accepted",
		"exp":   4000000000,
	})

	allowFallback := false
	_, err := VerifyPviumWebhookToken(token, secret, &models.VerifyPviumWebhookTokenOptions{
		AllowHashedSecretFallback: &allowFallback,
	})
	if err == nil {
		t.Fatal("expected verification to fail when hashed-secret fallback is disabled")
	}
}

func TestResolvePviumWebhookPayloadWithoutTokenUsesBodyEventAndData(t *testing.T) {
	t.Parallel()

	payload, err := ResolvePviumWebhookPayload(map[string]any{
		"event": "invoice.created",
		"data":  map[string]any{"id": "evt_1"},
	}, "unused-secret")
	if err != nil {
		t.Fatalf("resolve tokenless webhook payload: %v", err)
	}
	if payload.Event != "invoice.created" {
		t.Fatalf("event mismatch: got %s want %s", payload.Event, "invoice.created")
	}
	if payload.Data["id"] != "evt_1" {
		t.Fatalf("data id mismatch: got %v want %s", payload.Data["id"], "evt_1")
	}
}

File: node-sdk/test/webhooks.test.js
const test = require("node:test");
const assert = require("node:assert/strict");
const { createHmac, createHash } = require("node:crypto");

const {
  resolvePviumWebhookPayload,
  verifyPviumWebhookToken,
} = require("../dist/index.js");

function signJwt(payload, secret) {
  const encodedHeader = base64UrlJson({ alg: "HS256", typ: "JWT" });
  const encodedPayload = base64UrlJson(payload);
  const signature = createHmac("sha256", secret)
    .update(`${encodedHeader}.${encodedPayload}`)
    .digest("base64url");

  return `${encodedHeader}.${encodedPayload}.${signature}`;
}

function base64UrlJson(value) {
  return Buffer.from(JSON.stringify(value)).toString("base64url");
}

test("verifyPviumWebhookToken verifies backend HS256 webhook JWTs", () => {
  const token = signJwt(
    {
      event: "oauth.invite.accepted",
      data: { githubLogin: "octocat" },
      iat: 1_700_000_000,
      exp: 4_000_000_000,
    },
    "webhook_secret",
  );

  const payload = verifyPviumWebhookToken(token, "webhook_secret", {
    expectedEvent: "oauth.invite.accepted",
  });

  assert.equal(payload.event, "oauth.invite.accepted");
  assert.deepEqual(payload.data, { githubLogin: "octocat" });
});

test("verifyPviumWebhookToken supports backend tokens signed with sha256(secret)", () => {
  const secret = "secret_abc123";
  const hashedSecret = createHash("sha256").update(secret).digest("hex");
  const token = signJwt(
    {
      event: "batch.payee.added",
      data: { batch: { id: "batch_123" } },
      exp: 4_000_000_000,
    },
    hashedSecret,
  );

  const payload = verifyPviumWebhookToken(token, secret);

  assert.equal(payload.event, "batch.payee.added");
  assert.deepEqual(payload.data, { batch: { id: "batch_123" } });
});

test("resolvePviumWebhookPayload returns token data and checks body event", () => {
  const token = signJwt(
    {
      event: "invoice.paid",
      data: { invoiceId: "inv_123" },
      exp: 4_000_000_000,
    },
    "webhook_secret",
  );

  const resolved = resolvePviumWebhookPayload(
    { event: "invoice.paid", token },
    "webhook_secret",
  );

  assert.equal(resolved.event, "invoice.paid");
  assert.deepEqual(resolved.data, { invoiceId: "inv_123" });
});

test("verifyPviumWebhookToken rejects expired tokens", () => {
  const token = signJwt(
    {
      event: "invoice.paid",
      data: {},
      exp: 1_700_000_000,
    },
    "webhook_secret",
  );

  assert.throws(
    () =>
      verifyPviumWebhookToken(token, "webhook_secret", {
        now: 1_800_000_000_000,
      }),
    /Expired Pvium webhook token/,
  );
});
